frontend/src/liboperator/operator/devices.go

```diff
@@ -93,12 +107,41 @@ func (d *Device) ToClient(id int, msg interface{}) error {
 type DevicePool struct {
 	devicesMtx sync.Mutex
 	devices    map[string]*Device
+	groups     map[string]*Group
 }
 
 func NewDevicePool() *DevicePool {
 	return &DevicePool{
 		devices: make(map[string]*Device),
+		groups:  make(map[string]*Group),
+	}
+}
+
+func GetGroupId(info interface{}) string {
+	deviceInfo, ok := info.(map[string]interface{})
+	if !ok {
+		return DEFAULT_GROUP_ID
+	}
+
+	groupId, ok := deviceInfo["group_id"].(string)
+	if !ok {
+		return DEFAULT_GROUP_ID
+	}
+
+	if len(groupId) == 0 {
+		return DEFAULT_GROUP_ID
+	}
+
+	return groupId
+}
+
+func GetGroup(p *DevicePool, groupId string) *Group {
+	_, ok := p.groups[groupId]
```

Comment:
is it okay to return nil if the groupId is not found? if so, we can just return p.groups[groupId]